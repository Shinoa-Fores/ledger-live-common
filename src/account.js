/**
 * @flow
 * @module account
 */
import invariant from "invariant";
import { BigNumber } from "bignumber.js";
import type {
  Account,
  AccountIdParams,
  Operation,
  BalanceHistory,
  DailyOperations,
  CryptoCurrency,
  DerivationMode
} from "./types";
import { getOperationAmountNumber } from "./operation";
import {
  isSegwitDerivationMode,
  isUnsplitDerivationMode,
  asDerivationMode
} from "./derivation";

function startOfDay(t) {
  return new Date(t.getFullYear(), t.getMonth(), t.getDate());
}

/**
 * generate an array of {daysCount} datapoints, one per day,
 * for the balance history of an account.
 * The last item of the array is the balance available right now.
 * @memberof account
 */
export function getBalanceHistory(
  account: Account,
  daysCount: number
): BalanceHistory {
  const history = [];
  let { balance } = account;
  const operationsLength = account.operations.length;
  let i = 0; // index of operation
  let t = new Date();
  history.unshift({ date: t, value: balance });
  t = new Date(startOfDay(t) - 1); // end of yesterday
  for (let d = daysCount - 1; d > 0; d--) {
    // accumulate operations after time t
    while (i < operationsLength && account.operations[i].date > t) {
      balance = balance.minus(getOperationAmountNumber(account.operations[i]));
      i++;
    }
    if (i === operationsLength) {
      // When there is no more operation, we consider we reached ZERO to avoid invalid assumption that balance was already available.
      balance = BigNumber(0);
    }
    history.unshift({ date: t, value: BigNumber.max(balance, 0) });
    t = new Date(t - 24 * 60 * 60 * 1000);
  }
  return history;
}

/**
 * calculate the total balance history for all accounts in a reference fiat unit
 * and using a CalculateCounterValue function (see countervalue helper)
 * NB the last item of the array is actually the current total balance.
 * @memberof account
 */
export function getBalanceHistorySum(
  accounts: Account[],
  daysCount: number,
  // for a given account, calculate the countervalue of a value at given date.
  calculateAccountCounterValue: (Account, BigNumber, Date) => BigNumber
): BalanceHistory {
  if (typeof calculateAccountCounterValue !== "function") {
    throw new Error(
      "getBalanceHistorySum signature has changed, please port the code"
    );
  }
  if (accounts.length === 0) {
    const now = Date.now();
    const zeros: BigNumber[] = Array(daysCount).fill(BigNumber(0));
    return zeros.map((value, i) => ({
      date: new Date(now - 24 * 60 * 60 * 1000 * (daysCount - i - 1)),
      value
    }));
  }
  return accounts
    .map(account => {
      const history = getBalanceHistory(account, daysCount);
      return history.map(h => ({
        date: h.date,
        value: calculateAccountCounterValue(account, h.value, h.date)
      }));
    })
    .reduce((acc, history) =>
      acc.map((a, i) => ({
        date: a.date,
        value: a.value.plus(history[i].value)
      }))
    );
}

const emptyDailyOperations = { sections: [], completed: true };

/**
 * @memberof account
 */
export function groupAccountsOperationsByDay(
  accounts: Account[],
  count: number
): DailyOperations {
  // Track indexes of account.operations[] for each account
  const indexes: number[] = Array(accounts.length).fill(0);
  // Track indexes of account.pendingOperations[] for each account
  const indexesPending: number[] = Array(accounts.length).fill(0);
  // Returns the most recent operation from the account with current indexes
  function getNextOperation(): ?Operation {
    let bestOp: ?Operation;
    let bestOpInfo = { accountI: 0, fromPending: false };
    for (let i = 0; i < accounts.length; i++) {
      // look in operations
      const op = accounts[i].operations[indexes[i]];
      if (op && (!bestOp || op.date > bestOp.date)) {
        bestOp = op;
        bestOpInfo = { accountI: i, fromPending: false };
      }
      // look in pending operations
      const opP = accounts[i].pendingOperations[indexesPending[i]];
      if (opP && (!bestOp || opP.date > bestOp.date)) {
        bestOp = opP;
        bestOpInfo = { accountI: i, fromPending: true };
      }
    }
    if (bestOp) {
      if (bestOpInfo.fromPending) {
        indexesPending[bestOpInfo.accountI]++;
      } else {
        indexes[bestOpInfo.accountI]++;
      }
    }
    return bestOp;
  }

  let op = getNextOperation();
  if (!op) return emptyDailyOperations;
  const sections = [];
  let day = startOfDay(op.date);
  let data = [];
  for (let i = 0; i < count && op; i++) {
    if (op.date < day) {
      sections.push({ day, data });
      day = startOfDay(op.date);
      data = [op];
    } else {
      data.push(op);
    }
    op = getNextOperation();
  }
  sections.push({ day, data });
  return {
    sections,
    completed: !op
  };
}

/**
 * Return a list of `{count}` operations grouped by day.
 * @memberof account
 */
export function groupAccountOperationsByDay(
  account: Account,
  count: number
): DailyOperations {
  return groupAccountsOperationsByDay([account], count);
}

export function encodeAccountId({
  type,
  version,
  currencyId,
  xpubOrAddress,
  derivationMode
}: AccountIdParams) {
  return `${type}:${version}:${currencyId}:${xpubOrAddress}:${derivationMode}`;
}

export function decodeAccountId(accountId: string): AccountIdParams {
  invariant(typeof accountId === "string", "accountId is not a string");
  const splitted = accountId.split(":");
  invariant(splitted.length === 5, "invalid size for accountId");
  const [type, version, currencyId, xpubOrAddress, derivationMode] = splitted;
  return {
    type,
    version,
    currencyId,
    xpubOrAddress,
    derivationMode: asDerivationMode(derivationMode)
  };
}

// you can pass account because type is shape of Account
// wallet name is a lib-core concept that usually identify a pool of accounts with the same (seed, cointype, derivation scheme) config.
export function getWalletName({
  seedIdentifier,
  derivationMode,
  currency
}: {
  seedIdentifier: string,
  derivationMode: DerivationMode,
  currency: CryptoCurrency
}) {
  return `${seedIdentifier}_${currency.id}_${derivationMode}`;
}

const MAX_ACCOUNT_NAME_SIZE = 50;

const alwaysSuffixLegacyExcept = ["", "segwit", "segwit_unsplit", "unsplit"];

export const getAccountPlaceholderName = ({
  currency,
  index,
  derivationMode
}: {
  currency: CryptoCurrency,
  index: number,
  derivationMode: DerivationMode
}) =>
  `${currency.name} ${index + 1}${
    (!isSegwitDerivationMode(derivationMode) && currency.supportsSegwit) ||
    !alwaysSuffixLegacyExcept.includes(derivationMode)
      ? " (legacy)"
      : ""
  }${isUnsplitDerivationMode(derivationMode) ? " (unsplit)" : ""}`;

// An account is empty if there is no operations AND balance is zero.
// balance can be non-zero in edgecases, for instance:
// - Ethereum contract only funds (api limitations)
// - Ripple node that don't show all ledgers and if you have very old txs

export const isAccountEmpty = (a: Account): boolean =>
  a.operations.length === 0 && a.balance.isZero();

export const getNewAccountPlaceholderName = getAccountPlaceholderName; // same naming

export const validateNameEdition = (account: Account, name: ?string): string =>
  (
    (name || account.name || "").replace(/\s+/g, " ").trim() ||
    account.name ||
    getAccountPlaceholderName(account)
  ).slice(0, MAX_ACCOUNT_NAME_SIZE);

export type SortAccountsParam = {
  accounts: Account[],
  accountsBtcBalance: BigNumber[],
  orderAccounts: string
};

type SortMethod = "name" | "balance";

const sortMethod: { [_: SortMethod]: (SortAccountsParam) => string[] } = {
  balance: ({ accounts, accountsBtcBalance }) =>
    accounts
      .map((a, i) => [a.id, accountsBtcBalance[i] || BigNumber(-1), a.name])
      .sort((a, b) => {
        const numOrder = a[1].minus(b[1]).toNumber();
        if (numOrder === 0) {
          return a[2].localeCompare(b[2]);
        }

        return numOrder;
      })
      .map(o => o[0]),

  name: ({ accounts }) =>
    accounts
      .slice(0)
      .sort((a, b) => a.name.localeCompare(b.name))
      .map(a => a.id)
};

export function sortAccounts(param: SortAccountsParam) {
  const [order, sort] = param.orderAccounts.split("|");
  if (order === "name" || order === "balance") {
    const ids = sortMethod[order](param);
    if (sort === "desc") {
      ids.reverse();
    }
    return ids;
  }
  return null;
}
