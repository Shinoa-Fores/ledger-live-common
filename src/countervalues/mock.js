// @flow
import type { CounterValuesState, Histodays } from "./types";
import type { Currency } from "../types";
import { formatCounterValueDay } from ".";

type Pair = {
  from: Currency,
  to: Currency,
  exchange: string,
  dateFrom: Date,
  dateTo: Date,
  rate: Date => number
};

export const genDateRange = ({ dateFrom, dateTo, rate }: Pair): Histodays => {
  const histodays = {};
  for (let d = dateFrom; d < dateTo; d += 24 * 60 * 60 * 1000) {
    const day = new Date(d);
    histodays[formatCounterValueDay(day)] = rate(new Date(d));
  }
  histodays.latest = rate(new Date());
  return histodays;
};

export const genStoreState = (pairs: Pair[]): CounterValuesState => {
  const state = { rates: {} };
  pairs.forEach(pair => {
    const { from, to, exchange } = pair;
    const a = (state.rates[to.ticker] = state.rates[to.ticker] || {});
    const b = (a[from.ticker] = a[from.ticker] || {});
    b[exchange] = genDateRange(pair);
  });
  return state;
};
