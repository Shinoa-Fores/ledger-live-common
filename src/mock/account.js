/**
 * @module mock/account
 * @flow
 */
import { BigNumber } from "bignumber.js";
import { listCryptoCurrencies } from "../currencies";
import Prando from "prando";
import type { Account, Operation, CryptoCurrency } from "../types";
import { getOperationAmountNumber } from "../operation";
import { getDerivationScheme, runDerivationScheme } from "../derivation";

// for the mock generation we need to adjust to the actual market price of things, we want to avoid having things < 0.01 EUR
const tickerApproxMarketPrice = {
  BTC: 0.0073059,
  ETH: 5.7033e-14,
  ETC: 1.4857e-15,
  BCH: 0.0011739,
  BTG: 0.00005004,
  LTC: 0.00011728,
  XRP: 0.000057633,
  DOGE: 4.9e-9,
  DASH: 0.0003367,
  PPC: 0.000226,
  ZEC: 0.000205798
};

// mock only use subset of cryptocurrencies to not affect tests when adding coins
const currencies = listCryptoCurrencies().filter(
  c => tickerApproxMarketPrice[c.ticker]
);

/**
 * @memberof mock/account
 */
export function genBitcoinAddressLike(rng: Prando) {
  const charset = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  return `1${rng.nextString(rng.nextInt(25, 34), charset)}`;
}

/**
 * @memberof mock/account
 */
export function genHex(length: number, rng: Prando) {
  return rng.nextString(length, "0123456789ABCDEF");
}

/**
 * @memberof mock/account
 */
export function genAddress(currency: CryptoCurrency, rng: Prando) {
  if (currency.id === "ethereum" || currency.id === "ethereum_classic") {
    return `0x${genHex(40, rng)}`;
  }
  return genBitcoinAddressLike(rng);
}

// TODO fix the mock to never generate negative balance...
/**
 * @memberof mock/account
 */
export function genOperation(
  account: Account,
  ops: *,
  currency: CryptoCurrency,
  rng: Prando
): $Exact<Operation> {
  const lastOp = ops[ops.length - 1];
  const date = new Date(
    (lastOp ? lastOp.date : Date.now()) -
      rng.nextInt(0, 100000000 * rng.next() * rng.next())
  );
  const address = genAddress(currency, rng);
  const type = rng.next() < 0.3 ? "OUT" : "IN";
  const value = BigNumber(
    Math.floor(
      rng.nextInt(0, 100000 * rng.next() * rng.next()) /
        (tickerApproxMarketPrice[currency.ticker] ||
          tickerApproxMarketPrice.BTC)
    )
  );
  if (isNaN(value)) {
    throw new Error("invalid amount generated for " + currency.id);
  }
  return {
    id: String(`mock_op_${ops.length}_${account.id}`),
    hash: genHex(64, rng),
    type,
    value,
    fee: BigNumber(Math.round(value.toNumber() * 0.01)),
    senders: [type !== "IN" ? genAddress(currency, rng) : address],
    recipients: [type === "IN" ? genAddress(currency, rng) : address],
    blockHash: genHex(64, rng),
    blockHeight: account.blockHeight - Math.floor((Date.now() - date) / 900000),
    accountId: account.id,
    date,
    extra: {}
  };
}

/**
 * @memberof mock/account
 */
export function genAddingOperationsInAccount(
  account: Account,
  count: number,
  seed: number | string
): Account {
  const rng = new Prando(seed);
  const copy: Account = { ...account };
  copy.operations = Array(count)
    .fill(null)
    .reduce(ops => {
      const op = genOperation(copy, ops, copy.currency, rng);
      return ops.concat(op);
    }, copy.operations);
  copy.balance = ensureNoNegative(copy.operations);
  return copy;
}

/**
 * @param id is a number or a string, used as an account identifier and as a seed for the generation.
 * @memberof mock/account
 */
type GenAccountOptions = {
  operationsSize?: number,
  currency?: CryptoCurrency
};

export function genAccount(
  id: number | string,
  opts: GenAccountOptions = {}
): $Exact<Account> {
  const rng = new Prando(id);
  const currency = opts.currency || rng.nextArrayItem(currencies);
  const operationsSize = opts.operationsSize || rng.nextInt(1, 200);
  const address = genAddress(currency, rng);
  const account = {
    id: `mock:1:${currency.id}:${id}:`,
    seedIdentifier: "mock",
    derivationMode: "",
    xpub: genHex(64, rng),
    index: 1,
    freshAddress: address,
    freshAddressPath: runDerivationScheme(
      getDerivationScheme({ currency, derivationMode: "" }),
      currency
    ),
    name: rng.nextString(rng.nextInt(4, 34)),
    balance: BigNumber(0),
    blockHeight: rng.nextInt(100000, 200000),
    currency,
    unit: rng.nextArrayItem(currency.units),
    operations: [],
    pendingOperations: [],
    lastSyncDate: new Date()
  };

  account.operations = Array(operationsSize)
    .fill(null)
    .reduce((ops: Operation[]) => {
      const op = genOperation(account, ops, currency, rng);
      return ops.concat(op);
    }, []);

  account.balance = ensureNoNegative(account.operations);
  return account;
}

function ensureNoNegative(operations) {
  let total = BigNumber(0);
  for (let i = operations.length - 1; i >= 0; i--) {
    const op = operations[i];
    const amount = getOperationAmountNumber(op);
    if (total.plus(amount).isNegative()) {
      if (op.type === "IN") {
        op.type = "OUT";
      } else if (op.type === "OUT") {
        op.type = "IN";
      }
    }
    total = total.plus(getOperationAmountNumber(op));
  }
  return total;
}
