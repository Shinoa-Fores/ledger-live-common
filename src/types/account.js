// @flow

import type { BigNumber } from "bignumber.js";
import type { CryptoCurrency, Unit } from "./currencies";
import type { OperationRaw, Operation } from "./operation";
import type { DerivationMode } from "../derivation";

export type Account = {
  // unique account identifier
  id: string,

  // a unique way to identify a seed the account was associated with
  // it MUST be different between 2 seeds
  // but it is not necessarily the same between 2 accounts (if possible â€“ not always possible)
  // in BTC like accounts, we use pubKey(purpose'/coinType')
  // For other accounts that don't have sub derivation, we have used the account address
  seedIdentifier: string,

  // account xpub if available
  xpub?: string,

  // Identify the derivation used. it allows us to map this to a derivation scheme.
  // exemple of values: segwit | unsplit | segwit_unsplit | mew | eth_mew (eg for etc accounts on eth)
  // the special value of '' means it's bip44 with purpose 44.
  derivationMode: DerivationMode,

  // the iterated number to derivate the account in a given derivationMode config
  // in context of bip44, it would be the account field of bip44 ( m/purpose'/cointype'/account' )
  index: number,

  // next receive address. to be used to display to user.
  freshAddress: string,

  // The path linked to freshAddress. to be used to validate with the device if it corresponds to freshAddress.
  // exemple: 44'/0'/0'/0/0
  freshAddressPath: string,

  // account name
  name: string,

  // account balance in satoshi
  balance: BigNumber,

  // the last block height currently synchronized
  blockHeight: number,

  // ------------------------------------- Specific account fields

  // currency of this account
  currency: CryptoCurrency,

  // user preferred unit to use. unit is coming from currency.units. You can assume currency.units.indexOf(unit) will work. (make sure to preserve reference)
  unit: Unit,

  // lazy list of operations that exists on the blockchain.
  operations: Operation[],

  // pending operations that has been broadcasted but are not yet in operations
  // this is for optimistic updates UI. the Operation objects are temporary and
  // might not be the real one that will arrives on operations array.
  // only Operation#id needs to be guaranteed the same.
  // the array resulting of pendingOperations.concat(operations)
  // is guaranteed to contains unique ops (by id) at any time and also is time DESC sorted.
  pendingOperations: Operation[],

  // used to know when the last sync happened
  lastSyncDate: Date,

  // A configuration for the endpoint to use. (usecase: Ripple node)
  endpointConfig?: ?string
};

export type AccountRaw = {
  id: string,
  seedIdentifier: string,
  xpub?: string,
  derivationMode: DerivationMode,
  index: number,
  freshAddress: string,
  freshAddressPath: string,
  name: string,
  balance: string,
  blockHeight: number,
  // ------------------------------------- Specific raw fields
  currencyId: string,
  operations: OperationRaw[],
  pendingOperations: OperationRaw[],
  unitMagnitude: number,
  lastSyncDate: string,
  endpointConfig?: ?string
};
