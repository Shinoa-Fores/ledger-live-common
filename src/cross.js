// @flow
// cross helps dealing with cross-project feature like export/import & cross project conversions

import { BigNumber } from "bignumber.js";
import compressjs from "compressjs";
import type { Account, CryptoCurrencyIds, DerivationMode } from "./types";
import { runDerivationScheme, getDerivationScheme } from "./derivation";
import { decodeAccountId } from "./account";
import { getCryptoCurrencyById } from "./currencies";

export type AccountData = {
  id: string,
  currencyId: string,
  seedIdentifier: string,
  derivationMode: DerivationMode,
  name: string,
  index: number,
  balance: string
};

export type CryptoSettings = {
  exchange?: ?string,
  confirmationsNb?: number
};

export type Settings = {
  counterValue?: string,
  counterValueExchange?: ?string,
  currenciesSettings: {
    [_: CryptoCurrencyIds]: CryptoSettings
  }
};

export type DataIn = {
  // accounts to export (filter them to only be the visible ones)
  accounts: Account[],
  // settings
  settings: Settings,
  // the name of the exporter. e.g. "desktop" for the desktop app
  exporterName: string,
  // the version of the exporter. e.g. the desktop app version
  exporterVersion: string
};

export type Result = {
  accounts: AccountData[],
  settings: Settings,
  meta: {
    exporterName: string,
    exporterVersion: string
  }
};

export function encode({
  accounts,
  settings,
  exporterName,
  exporterVersion
}: DataIn): string {
  return Buffer.from(
    compressjs.Bzip2.compressFile(
      Buffer.from(
        JSON.stringify({
          meta: { exporterName, exporterVersion },
          accounts: accounts.map(accountToAccountData),
          settings
        })
      )
    )
  ).toString("binary");
}

export function decode(bytes: string): Result {
  return JSON.parse(
    Buffer.from(
      compressjs.Bzip2.decompressFile(Buffer.from(bytes, "binary"))
    ).toString()
  );
}

export function accountToAccountData({
  id,
  name,
  seedIdentifier,
  derivationMode,
  currency,
  index,
  balance
}: Account): AccountData {
  return {
    id,
    name,
    seedIdentifier,
    derivationMode,
    currencyId: currency.id,
    index,
    balance: balance.toString()
  };
}

// reverse the account data to an account.
// this restore the essential data of an account and the result of the fields
// are assumed to be restored during first sync
export const accountDataToAccount = ({
  id,
  currencyId,
  name,
  index,
  balance,
  derivationMode,
  seedIdentifier
}: AccountData): Account => {
  const { xpubOrAddress } = decodeAccountId(id); // TODO rename in AccountId xpubOrAddress
  const currency = getCryptoCurrencyById(currencyId);
  let xpub = "";
  let freshAddress = "";
  let freshAddressPath = "";
  if (currency.family === "bitcoin") {
    xpub = xpubOrAddress;
  } else {
    freshAddress = xpubOrAddress;
    freshAddressPath = runDerivationScheme(
      getDerivationScheme({ currency, derivationMode }),
      currency,
      { account: index }
    );
  }

  const account: $Exact<Account> = {
    id,
    derivationMode,
    seedIdentifier,
    xpub,
    name,
    currency,
    index,
    freshAddress,
    freshAddressPath,
    // these fields will be completed as we will sync
    blockHeight: 0,
    balance: BigNumber(balance),
    operations: [],
    pendingOperations: [],
    unit: currency.units[0],
    lastSyncDate: new Date(0)
  };
  return account;
};
