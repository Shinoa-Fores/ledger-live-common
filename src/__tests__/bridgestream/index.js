// @flow
import { encode, decode } from "../../cross";
import shuffle from "lodash/shuffle";
import { genAccount } from "../../mock/account";

test("encode/decode", () => {
  const accounts = Array(3)
    .fill(null)
    .map((_, i) => genAccount("export_" + i));
  const arg = {
    accounts,
    settings: {
      counterValue: "USD",
      counterValueExchange: "KRAKEN",
      currenciesSettings: {
        bitcoin: {
          exchange: "KRAKEN"
        }
      }
    },
    exporterName: "test",
    exporterVersion: "0.0.0",
    chunkSize: 100
  };
  const data = encode(arg);
  const res = decode(data);
  expect(res.accounts).toMatchObject(
    accounts.map(a => ({
      balance: a.balance.toString(),
      currencyId: a.currency.id,
      id: a.id,
      name: a.name,
      index: a.index
    }))
  );
  expect(res.settings).toMatchObject({
    counterValue: "USD",
    counterValueExchange: "KRAKEN",
    currenciesSettings: {
      bitcoin: {
        exchange: "KRAKEN"
      }
    }
  });
  expect(res).toMatchSnapshot();
});
