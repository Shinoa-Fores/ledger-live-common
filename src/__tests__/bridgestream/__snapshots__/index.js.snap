// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode/decode 1`] = `
Object {
  "accounts": Array [
    Object {
      "balance": "1797257488",
      "currencyId": "zcash",
      "derivationMode": "",
      "id": "mock:1:zcash:export_0:",
      "index": 1,
      "name": "PcbUsj9a5mnpn6wT6InSD9n",
      "seedIdentifier": "mock",
    },
    Object {
      "balance": "1372345691",
      "currencyId": "zcash",
      "derivationMode": "",
      "id": "mock:1:zcash:export_1:",
      "index": 1,
      "name": "rv6RjflCC8",
      "seedIdentifier": "mock",
    },
    Object {
      "balance": "2862301845",
      "currencyId": "zcash",
      "derivationMode": "",
      "id": "mock:1:zcash:export_2:",
      "index": 1,
      "name": "f9mx3pw3IAcuP7ovSPao6EQ",
      "seedIdentifier": "mock",
    },
  ],
  "meta": Object {
    "exporterName": "test",
    "exporterVersion": "0.0.0",
  },
  "settings": Object {
    "counterValue": "USD",
    "counterValueExchange": "KRAKEN",
    "currenciesSettings": Object {
      "bitcoin": Object {
        "exchange": "KRAKEN",
      },
    },
  },
}
`;
